src/sidebar/components/ThreadList.tsx

```diff
@@ -33,6 +34,45 @@ export type ThreadListProps = {
   threads: Thread[];
 };
 
+/**
+ * Find the selector identifying the document segment which an annotation
+ * belongs to.
+ */
+function getSegmentSelector(ann: Annotation): EPUBContentSelector | undefined {
+  return ann.target[0].selector?.find(s => s.type === 'EPUBContentSelector') as
+    | EPUBContentSelector
+    | undefined;
+}
+
+/**
+ * Return a key that identifies the document section to which an annotation
+ * belongs.
+ */
+function headingKey(thread: Thread): string | null {
+  if (!thread.annotation) {
+    return null;
+  }
+  const chapter = getSegmentSelector(thread.annotation);
+  return chapter?.cfi ?? null;
+}
+
+/** Build a map of heading key (see {@link headingKey}) to section heading. */
+function headingMap(threads: Thread[]): Map<string, string> {
+  const headings = new Map();
+
+  for (const thread of threads) {
+    if (!thread.annotation) {
+      continue;
+    }
+    const selector = getSegmentSelector(thread.annotation);
+    if (selector?.title) {
+      headings.set(selector.cfi, selector.title);
+    }
+  }
+
+  return headings;
+}
```

Comment:
If the logic here gets significantly more complex, I would extract it into a separate module.

Reply:
Yes, this feels a little out of place but I'm willing to go with it for now while this feature is being sorted through...